FluidNC/src/Maslow/MotorUnit.cpp

```diff
@@ -141,176 +104,156 @@ double MotorUnit::recomputePID(){
  */
-bool MotorUnit::comply(unsigned long *timeLastMoved, double *lastPosition, double *amtToMove, double maxSpeed){
-    
+bool MotorUnit::comply( double maxSpeed){
+
+    //Call it every 50 ms
+    if(millis() - lastCallToComply < 50){
+        return true;
+    }
     //Update position and PID loop
     recomputePID();
-    updateEncoderPosition();
 
     //If we've moved any, then drive the motor outwards to extend the belt
     float positionNow = getPosition();
-    float distMoved = positionNow - *lastPosition;
+    float distMoved = positionNow - lastPosition;
 
     //If the belt is moving out, let's keep it moving out
     if( distMoved > .001){
         //Increment the target
-        setTarget(positionNow + *amtToMove);
+        setTarget(positionNow + amtToMove);
         
-        *amtToMove = *amtToMove + 1;
+        amtToMove = amtToMove + 1;
         
-        *amtToMove = min(*amtToMove, maxSpeed);
+        amtToMove = min(amtToMove, maxSpeed);
         
         //Reset the last moved counter
-        *timeLastMoved = millis();
+        //*timeLastMoved = millis();
     
     //If the belt is moving in we need to stop it from moving in
     }else if(distMoved < -.04){
-        *amtToMove = 0;
+        amtToMove = 0;
         setTarget(positionNow + .1);
         stop();
     }
     //Finally if the belt is not moving we want to spool things down
     else{
-        *amtToMove = *amtToMove / 2;
+        amtToMove = amtToMove / 2;
         setTarget(positionNow);
         stop();
     }
     
 
-    *lastPosition = positionNow;
+    lastPosition = positionNow;
 
     //Return indicates if we have moved within the timeout threshold
-    if(millis()-*timeLastMoved > 5000){
-        return false;
-    }
-    else{
+    // if(millis()-*timeLastMoved > 5000){
+    //     return false;
+    // }
+    //else{
+        lastCallToComply = millis();
         return true;
-    }
+    //}
 }
 
 /*!
- *  @brief  Fully retracts this axis and zeros it out or if it is already retracted extends it to the targetLength
+ *  @brief  Fully retracts this axis and zeros it, non-blocking, returns true when done
  */
-bool MotorUnit::retract(double targetLength){
-    
-    log_info("Retracting");
-
-    int absoluteCurrentThreshold = 1900;
-    int incrementalThreshold = 75;
-    int incrementalThresholdHits = 0;
-    float alpha = .2;
-    float baseline = 700;
-
-    uint16_t speed = 0;
+bool MotorUnit::retract(){
 
-    //Keep track of the elapsed time
-    unsigned long time = millis();
-    unsigned long elapsedTime = 0;
-    
-    //Pull until taught
-    while(true){
-        
         //Gradually increase the pulling speed
-        if(random(0,4) == 2){ //This is a hack to make it speed up more slowly because we can't add less than 1 to an int
-            speed = min(speed + 1, 1023);
+        if(random(0,2) == 1){ //This is a hack to make it speed up more slowly because we can't add less than 1 to an int
```

Comment:
This will be twice as fast, not sure if that's the intention. I've been using this trick other places and I think I wrote it in an unclear way here (this way makes it seem like the 1 or 2 matters). It's really just a 1 in 4 coin toss. I think that random(2)==0 is more clear...although now I can't remember if that's 1 in 2 chance or a 1 in 3 chance...Are the possible outcomes 0,1 or 0,1,2?